### PR TITLE
Allow to pack a Windows app as zip file

### LINKED ISSUE #457: Create Windows executable instead of MSI
Hi, I have two questions: 
Is it possible to create a self-contained Windows executable file instead of an installable version? Basically I'm looking for something similar to how the packaging is done already for linux.
If it isn't at the moment, will it ever be possible? Thanks.

### PR DESCRIPTION
**Pull request by MarkusPiotrowski**
As discussed in issue #457 and elsewhere on the discord channel, a Windows app packed as zip file could be a nice alternative for shipping and will sometimes be preferable to an MSI installer.

I also got the impression that @freakboy3742 wouldn't block such a PR per se, so I'll give it a try ;-)

Fixes #457 

## PR Checklist:
- [X] All new features have been tested
- [X] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

### CODE REVIEW

File: changes/457.feature.rst
Comment (freakboy3742) [inline changes/457.feature.rst:1]:
```suggestion
Support for packaging Windows apps as simple Zip files was added.
```

File: docs/reference/platforms/windows/app.rst
Comment (freakboy3742) [inline docs/reference/platforms/windows/app.rst:17]: This should be incorporated into the earlier paragraph introducing WiX; something like:

```
Briefcase supports 2 packaging formats for a Windows app:

1. As an MSI installer
2. As a Zipfile containing all files needed to run the app
```
then lead into the description of how WiX is used to build MSIs.

A similar comment/update is needed on the VisualStudio format as well.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Thanks for the contribution! The broad strokes of this look good; I've left some comments inline about some of the specifics.

**Comment by MarkusPiotrowski** [comment]
Dear @freakboy3742 ,

Thank you very much for your detailed review and I'm sorry for confronting you with a PR with so much issues. One of the problems I had was that I had a running implementation on briefcase 0.3.12 (where app signing and distribution folders didn't exist) but then I was pushed back with each new release of briefcase since they had some significant changes for Windows packaging. Most of the time for this PR I spend trying to understand the testing environment (and obviously I did not well), for two days I believed that my implementation is not longer working because the tests failed until I realized that the implementation was OK and the issues were 'only' the tests... :-)
I hope I can address all the issues you mentioned, I'll try it.

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:35]: I guess you're right. Regarding creating a new fixture that extends an existing one I failed. Fixtures cannot be called directly and I didn't find the way to do so.

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:267]: I was thinking about a root folder, wasn't sure if this is wanted. The path issue origins from the fact that the distribution path as defined earlier in the code does already contain the `.msi` suffix. 
After writing the shutil command from scratch I finally realized that zip packaging is also used in the Web backend, so I took it from there.

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:239]: It does not only give you coverage, it will also show you that creating a zip folder with the correct name does work.
My first guess was that within the testing environment I find some real hello world project that is used for building, running, packaging and which I can use. When I realized that e.g. the msi packager tests does not actually create an MSI installer, I wondered how important testing the zip content is considered crucial here. Of course, I can put the correct folder structure in the zip and add some more mock files, still, this wouldn't show that the packed app will run.

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:35]: Got it.

File: tests/platforms/conftest.py
Comment (MarkusPiotrowski) [reply tests/platforms/conftest.py:21]: Made a local `conftest.py` file in the platform/windows file.

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [reply src/briefcase/platforms/windows/__init__.py:267]: The web backend also does zip packaging, but the use case there explicitly *doesn't* need the root folder. You unpack a web project into the folder where you want it to be served. 

As for the issue with `distribution_path` already containing `msi` - that's my point - it shouldn't contain `msi` if you're packaging a zip file. At the very least, it's returning the wrong extension. `distribution_path` should be adapting to the selected packaging format.

**Comment by freakboy3742** [comment]
> Thank you very much for your detailed review and I'm sorry for confronting you with a PR with so much issues. 

Not a problem at all - and although there's a lot of comments, I don't consider the issues to be *that* bad - it's lots of little things that are, in many cases, wrapped up in details that you probably don't know about unless you've been soaking in the code for years like I have. This is a really good first-time project contribution - the fact that you've submitted a PR that fully passes CI on your first attempt (even if there's room for improving those tests) puts you at the top of the class in terms of contributions :-)

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:267]: For the moment, I see two options to make a root folder:
- creating a folder in the build folder (e.g. on the same level as src) and copy the content of the src folder into it (this I have implemented now). In this case, should we delete the root folder afterward or just keep it?
- temporarily rename the build/.../scr folder to `MyApp-1.2.3`. This seems a little bit 'hackish' to me.
What would you recommend? Unfortunately, neither `ZipFile` nor `shutil.make_archive` have a 'put in a root folder before zipping' option...

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:263]: Regarding app signing: Should the final zip file also be signed (if signing is requested), like the msi file?

File: docs/reference/platforms/windows/index.rst
Comment (freakboy3742) [inline docs/reference/platforms/windows/index.rst:10]:
```suggestion
Both output formats support packaging as a Microsoft Software Installer (MSI) or
```

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [inline src/briefcase/platforms/windows/__init__.py:290]: This will attempt to sign the .zip file, which won't work because .zip files don't allow for signing.

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [inline src/briefcase/platforms/windows/__init__.py:384]: This works, but it (a) results in a copy of the entire project, which might not be small, and (b) could be prone to stale files on the second build.

At the very least, the copied folder should be a one-use temporary folder, or deleted after use; however a better approach may be to *not* use `shutil`, and use zip file directly - that gives you much better control over the structure of the final zip file. For example, see [this StackOverflow](https://stackoverflow.com/a/32640292)  answer.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Getting really close - a couple more comments inline.

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:290]: Yes, I asked this in the conversation above, but then I didn't wait for your answer. Sorry. Google search results have let me believe that this is possible. Seems that it is not possible with Windows' SDK SignTool.

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:384]: I have considered some options. The 'problem' here is that we want to put the files in a root folder _that does not already exist_ before zipping. This is different to the situation described in the StackOverflow question/answer (that I also found). My impression was that neither `shutil.make_archive` nor `ZipFile` allows the creation/building of a root folder 'on the fly' while packing, so I was thinking about either to temporarily rename the `src` folder to `MyApp-1.2.3` (what I found 'hackish') or to make a plain copy and delete it afterwards (what I'm doing now).

OK, after reading this [StackOverflow answer](https://stackoverflow.com/questions/458436/adding-folders-to-a-zip-file-using-python), I see that it is possible.

**Comment by MarkusPiotrowski** [comment]
Possibly, there was an unrelated error during the Python 3.9 test on macOs (CI / Unit test).
Can someone please restart the tests?

**Comment by rmartin16** [comment]
hmm...`test_stuck_streamer` has been problematic before...I thought we had fixed it, though; seems another look is necessary. Thanks for the ping.

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [inline src/briefcase/platforms/windows/__init__.py:385]: OK, this `except` is missing coverage.

@freakboy3742 @rmartin16: I saw that testing creates several temporary folders, for each test one (?). So when I want to test a condition that a folder was already created (like here), I need to test this within the same `test_...()` function, otherwise I end up in a fresh test folder, is this correct?

Comment (rmartin16) [reply src/briefcase/platforms/windows/__init__.py:385]: > I saw that testing creates several temporary folders, for each test one (?).

Yes; Pytest provides the `tmp_path` fixture for each test such that it's a unique empty directory that's ultimately cleaned up by Pytest.

> So when I want to test a condition that a folder was already created (like here), I need to test this within the same `test_...()` function, otherwise I end up in a fresh test folder, is this correct?
> 
> Or I create the `dist` folder in the fixture from the beginning, so no need to run the `_package_zip()` two times.

I'd probably need to see a specific test example...but moreover, you shouldn't need to create the `dist_path` directory here; it is [already created](https://github.com/beeware/briefcase/blob/5c83de4a0967eef40b6145e42025fbe37d2ba540/src/briefcase/commands/package.py#L93) by the base `PackageCommand`.

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [inline src/briefcase/platforms/windows/__init__.py:389]: We generally prefer the `pathlib` form of these commands to the `os` form - so `source.glob("*/**")` in this case. Similar updates are needed for the other `os` calls in this block.

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [reply src/briefcase/platforms/windows/__init__.py:385]: The mkdir call definitely isn't needed. It will be [created by the base package command](https://github.com/beeware/briefcase/blob/main/src/briefcase/commands/package.py#L93). That doesn't guarantee it exists for your *test* - so you may need to create the dist folder as part a fixture or other test setup.

File: tests/platforms/windows/app/test_package.py
Comment (freakboy3742) [inline tests/platforms/windows/app/test_package.py:31]: This isn't needed - `tmp_path` is guaranteed to be consistent within a given test. You don't need to annotated it onto the command.

File: tests/platforms/windows/app/test_package.py
Comment (freakboy3742) [inline tests/platforms/windows/app/test_package.py:63]: Although this evidently works on Windows, it would be preferable to use Pathlib objects and make the path separator unambiguous.

File: tests/platforms/windows/app/test_package.py
Comment (freakboy3742) [inline tests/platforms/windows/app/test_package.py:68]: We've got a `create_file()` utility method in `tests/utils` for this purpose - it creates a file with known content, and ensures the directory exists.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A few more cleanups - but we're getting really close!

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:63]: I'd prefer to let it stand as it is here; this list is also used to check the content of the ZIP file, and the namelist of the ZIP file consists of strings in this format.
However, the entries will be converted to Pathlib objects when creating the file with `create_file()`, as you suggested.

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:31]: OK, got it. These pytest fixtures a still enigmatic to me.

File: src/briefcase/platforms/windows/__init__.py
Comment (MarkusPiotrowski) [reply src/briefcase/platforms/windows/__init__.py:389]: OK, did so.

File: tests/platforms/windows/app/test_package.py
Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py:68]: Used it now.

File: tests/platforms/windows/app/test_package.py
Comment (freakboy3742) [reply tests/platforms/windows/app/test_package.py:63]: But that's not what this part of the test is for. It's creating fixtures. It's *essential* that they are created with the correct, platform appropriate path separators. The test fixture should *not* also be providing the test validation data - the test should be explicitly specifying what it expects, not relying on the input from a fixture.

Comment (freakboy3742) [reply tests/platforms/windows/app/test_package.py:63]: This is especially important when what is being validated isn't in the same format as what is being put into the system in the first place.

Comment (MarkusPiotrowski) [reply tests/platforms/windows/app/test_package.py]: OK, changed it.

File: src/briefcase/platforms/windows/__init__.py
Comment (freakboy3742) [reply src/briefcase/platforms/windows/__init__.py:293]: This looks like the only outstanding item.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is almost ready to land - the only outstanding issue I can see is the "package signing" issue. Indent that block so it only applies to MSI packaging, and we should be good.

It would also be desirable to have a test of that case, confirming that when a ZIP package is signed, only the app binary is signed. It doesn't need to actually do the signing; mocking ` `sign_app` and confirming the calls made should be sufficient.

**Comment by MarkusPiotrowski** [comment]
@freakboy3742 Just to be sure that I got it right: At the moment, when a zip has been packaged, it will not go on for signing, because I put a `return` in the `if` clause (so principally I could also leave out the `else`). Indenting the last signing block makes it visual clearer that it only applies to the msi package, but then I can also remove the `return`, can't I?
I look what I can do regarding an `test_zip_file_with_signing`.

**Comment by freakboy3742** [comment]
> @freakboy3742 Just to be sure that I got it right: At the moment, when a zip has been packaged, it will not go on for signing, because I put a `return` in the `if` clause (so principally I could also leave out the `else`). Indenting the last signing block makes it visual clearer that it only applies to the msi package, but then I can also remove the `return`, can't I?

My apologies - I completely missed the return in my scan of the code. 

As a general rule, we prefer to avoid using `return` as a flow control statement - "methods always enter at the top, and exit at the bottom" is a good expectation to set. There are some exceptions to this - usually when you need to shortcut logic due to an error or "no action required" case - but it's far too easy to accidentally miss an inline `return` statement (as I did in this case), so it's better to avoid the problem entirely. So - while the end result will be identical, the "indent" approach you've described would be preferred.

**Comment by rmartin16** [comment]
If I may, a small nitpick while we're here: the `description` for `WindowsVisualStudioPackageCommand` and `WindowsAppPackageCommand` both explicitly reference `MSI` as the package format. There's some variation in how other platforms word this, but updating these to just `Package a Visual Studio project.` and `Package a Windows app.`, respectively, probably makes sense.

**Comment by MarkusPiotrowski** [comment]
OK,  CI / Unit tests on macOS (for now: 3.8 and 3.9) throw a number of DeprecationWarnings.

**Comment by freakboy3742** [comment]
> OK, CI / Unit tests on macOS (for now: 3.8 and 3.9) throw a number of DeprecationWarnings. This is not the well known `test_stuck_streamer` issue : `DeprecationWarning: pkg_resources is deprecated as an API`

Oh wow - you're having the most remarkable luck hitting CI problems that aren't your fault at all :-)

This one is because there's been a recent update to setuptools that is causing dmgbuild, the tool that we use to build macOS DMG installers, to raise a warning that pytest is considering a failure. I know how to fix it, but I'll need to push out a new dmgbuild release and then update briefcase to use that new version.

**Comment by MarkusPiotrowski** [comment]
It' `test_stuck_streamer`-issue time again 🤪

**Comment by rmartin16** [comment]
I've opened #1197 to hopefully resolve the issues with `test_stuck_streamer`. #1196 should resolve the new `pip` issues once it's merged.

**Review by freakboy3742 — APPROVED** [review]
Awesome - Thanks for the contribution! The issue with the failing macOS test is unrelated to this patch, and CI is passing, so I'm happy to merge this, and deal with the intermittent test failure independently.

**Comment by MarkusPiotrowski** [comment]
Great! Thank you for your guidance with this PR.
Now I saw by chance, because of the recent discussion on discord about the briefcase templates, that there is an open ticket regarding leftover artifacts after the building process: https://github.com/beeware/briefcase-windows-app-template/issues/6#issue-1563430471
From that I learned that `python.exe` is an unwanted leftover, which is a pity, because that is one of the filenames that I put in the test module. Thus, I'm thinking about looking at this issue. My question: Except the two Python `.exe`s, is there other stuff that could/should be removed? If so, where could I learn about this?

**Comment by freakboy3742** [comment]
I'm not overly concerned that the specific files in the test case include `python.exe` - the list of files generated by the template is always going to change; the test case doesn't have to be a *perfect* match for reality - just a representative example of "files that might be packaged". 

As for whether there's other stuff that can be removed - I honestly don't know. I haven't done a full audit; the exes are the obvious ones, but there might be some additional files that aren't needed. In terms of how to learn about it... I'm not sure there's much by way of documentation. The best approach will be to try and work out what role each file in the archive actually serves, and build up a list of candidates for elimination.

**Pull request merged**
